### PR TITLE
535 include dispersion into the twiss parameters calculation and beam initialization using from twiss

### PR DESCRIPTION
**Pull request by cr-xu**
## Motivation and Context

- [ ] I have raised an issue to propose this change (required for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [ ] I have reformatted the code and checked that formatting passes (**required**).
- [ ] I have have fixed all issues found by `flake8` (**required**).
- [ ] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [ ] I have checked that the documentation builds (**required**).

Note: We are using a maximum length of 88 characters per line.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 - Implement second-order tracking for `Drift`, `Dipole` and `Quadrupole` elements, and add a convenient method to set tracking methods for an entire segment. This comes with an overhaul of the overall tracking system. Rename the tracking method `"cheetah"` to `"linear"` and `"bmadx"` to `"drift_kick_drift"`. The existing methods `"cheetah"` and `"bmadx"` will remain supported with a `DeprecationWarning`. (see #476) (@cr-xu, @jank324, @Hespe)
 - `Cavity` now supports travelling wave cavities in addition to standing wave cavities via the `cavity_type` argument (see #286) (@zihan-zh, @jank324)
 - Documented PyTorch `compile` for improved speed (see #390) (@ax3l)
+- Included dispersion correction into the twiss and emittance calculation in `Beam`, add dispersion terms in the `from_twiss` and `from_parameter` beam initialization methods (see #540) (@cr-xu)
```
Comment (jank324) [inline CHANGELOG.md:14]:
```suggestion
- Included dispersion correction into the Twiss and emittance calculation in `Beam`, add dispersion terms in the `from_twiss` and `from_parameter` beam initialization methods (see #540) (@cr-xu)
```

File: cheetah/particles/beam.py [hunk 2]
```diff
@@ -390,15 +390,57 @@ def cov_taup(self) -> torch.Tensor:
         raise NotImplementedError
 
     @property
-    def emittance_x(self) -> torch.Tensor:
-        """Emittance of the beam in x direction in m."""
+    @abstractmethod
+    def cov_xp(self) -> torch.Tensor:
+        """The covariance of (x, p)"""
```
Comment (jank324) [inline cheetah/particles/beam.py:395]:
```suggestion
        """Covariance of between `x` and `p`."""
```

File: cheetah/particles/beam.py [hunk 3]
```diff
@@ -390,15 +390,57 @@ def cov_taup(self) -> torch.Tensor:
...
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def cov_pxp(self) -> torch.Tensor:
+        """The covariance of (px, p)"""
```
Comment (jank324) [inline cheetah/particles/beam.py:401]:
```suggestion
        """Covariance between `px` and `p`."""
```

File: cheetah/particles/beam.py [hunk 4]
```diff
@@ -390,15 +390,57 @@ def cov_taup(self) -> torch.Tensor:
...
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def cov_yp(self) -> torch.Tensor:
+        """The covariance of (y, p)"""
```
Comment (jank324) [inline cheetah/particles/beam.py:407]:
```suggestion
        """Covariance between `y` and `p`."""
```

File: cheetah/particles/beam.py [hunk 5]
```diff
@@ -390,15 +390,57 @@ def cov_taup(self) -> torch.Tensor:
...
+    @abstractmethod
+    def cov_pyp(self) -> torch.Tensor:
+        """Covariance between `py` and `p`."""
+        raise NotImplementedError
+
+    @property
+    def projected_emittance_x(self) -> torch.Tensor:
+        """Projected emittance of the beam in x direction in m.
+        This is determined from the beam sizes without dispersion correction.
+        """
```
Comment (jank324) [inline cheetah/particles/beam.py:420]:
```suggestion
        """
        Projected emittance of the beam in x direction in m.
        This is determined from the beam sizes without dispersion correction.
        """
```

File: cheetah/particles/beam.py [hunk 6]
```diff
@@ -390,15 +390,57 @@ def cov_taup(self) -> torch.Tensor:
...
                 self.sigma_x**2 * self.sigma_px**2 - self.cov_xpx**2,
                 torch.finfo(self.sigma_x.dtype).tiny,
             )
         )
 
+    @property
+    def emittance_x(self) -> torch.Tensor:
+        """Uncoupled betatron emittance of the beam in x direction in m.
+        This is calculated with the dispersion correction.
+        """
```
Comment (jank324) [inline cheetah/particles/beam.py:432]:
```suggestion
        """
        Uncoupled betatron emittance of the beam in x direction in m.
        This is computed with the dispersion correction.
        """
```

File: cheetah/particles/beam.py [hunk 7]
```diff
@@ -407,23 +451,46 @@ def normalized_emittance_x(self) -> torch.Tensor:
     @property
     def beta_x(self) -> torch.Tensor:
         """Beta function in x direction in meters."""
-        return self.sigma_x**2 / self.emittance_x
+        sigma_p = torch.clamp_min(self.sigma_p, torch.finfo(self.sigma_x.dtype).tiny)
+        return (self.sigma_x**2 - self.cov_xp**2 / sigma_p**2) / self.emittance_x
```
Comment (jank324) [inline cheetah/particles/beam.py:455]: Do we really need the clamping here? (Also applies to all other clamping in this PR.) 

In #538, we are currently trying to get rid of all `clamp`s (and `if` and `where`) because they can be slow, and more critically break the compute graph for gradient computation.

Two ways they can usually be removed:
 - Rearranging equation
 - The case that the operation is trying to guard against (e.g. `sigma_p == 0.0`) would be unphysical, and therefore there is no need to catch it.

File: CHANGELOG.md [hunk 8]
```diff
@@ -11,6 +11,7 @@
 - Implement second-order tracking for `Drift`, `Dipole` and `Quadrupole` elements, and add a convenient method to set tracking methods for an entire segment. This comes with an overhaul of the overall tracking system. Rename the tracking method `"cheetah"` to `"linear"` and `"bmadx"` to `"drift_kick_drift"`. The existing methods `"cheetah"` and `"bmadx"` will remain supported with a `DeprecationWarning`. (see #476) (@cr-xu, @jank324, @Hespe)
 - `Cavity` now supports travelling wave cavities in addition to standing wave cavities via the `cavity_type` argument (see #286) (@zihan-zh, @jank324)
 - Documented PyTorch `compile` for improved speed (see #390) (@ax3l)
+- Included dispersion correction into the Twiss and emittance calculation in `Beam`, add dispersion terms in the `from_twiss` and `from_parameter` beam initialization methods (see #540) (@cr-xu)
```
Comment (jank324) [inline CHANGELOG.md:14]:
```suggestion
- Beam classes now account for dispersion. Dispersion correction is included in the Twiss and emittance computations. Dispersion arguments are added to `from_parameters` and `from_twiss` beam initialisation methods. (see #540) (@cr-xu)
```

**Review by jank324 — CHANGES_REQUESTED** [review]
Comment about clamping and 0.5 addition to energy in test.

File: cheetah/particles/beam.py (hunk 7 above)
Comment (cr-xu) [reply cheetah/particles/beam.py:471]: Good point. I agree that we should get rid of clamping.

File: tests/test_particle_beam.py [hunk 9]
```diff
@@ -246,7 +254,7 @@ def test_indexing_with_vectorized_incoming_beam():
     incoming = cheetah.ParticleBeam.from_parameters(
         num_particles=1_000,
         sigma_x=torch.tensor(1e-5),
-        energy=torch.rand((5, 2)) * 154e6,
+        energy=(0.5 + torch.rand((5, 2))) * 154e6,
```
[the remark this answers is not in the document]
Comment (cr-xu) [reply tests/test_particle_beam.py:257]: I remembered I encountered some sporadic failure which might be due to the very low energy randomly generated here. But I can't seem to reproduce it so I'll just revert the change.
